Config include exclude

ttaylorr opened the pull request:
@dipnlik reported #1251 the other day that the include/exclude filters made a distinction between `./foo/bar/` (with the trailing slash) and `./foo/bar` (without).

This pull-request introduces a patch that essentially runs filepath.Clean() against the paths before processing them. In addition, it removes the `command.determineIncludeExcludeFilters` function, and instead introduces a set of more granular replacement functions in the `config.Configuration` type. Those functions are all documented within the changeset.

ttaylorr commented:
As a side note, the `config.Configuration` type is quickly becoming a dumping ground for miscellaneous methods. I will try and spend this weekend thinking about alternative designs or patterns to refactor against, but am definitely more than open to suggestions on this as well.

For now, I think this is fine, as I assume it will be a pretty drawn-out process to preform this refactor, since so many parts of LFS depend on this type.

/cc @sinbad

sinbad commented on commands/commands.go:
```diff
@@ -210,30 +210,6 @@ type ErrorWithStack interface {
-// determineIncludeExcludePaths is a common function to take the string arguments
-// for include/exclude and derive slices either from these options or from the
-// common global config
-func determineIncludeExcludePaths(includeArg, excludeArg string) (include, exclude []string) {
```
I actually think this function was better where it was; it's specifically about converting string arguments to slices or defaulting and so belongs firmly unexported in `command` in my view. I don't see any benefit in promoting it to an exported function in `Configuration`, doing so widens that scope outside its remit IMO (which I think you were feeling too based on your follow up comment).

sinbad commented on config/config.go:
```diff
@@ -587,20 +588,68 @@ func (c *Configuration) readGitConfig(output string, uniqRemotes map[string]bool
+
+	return cleaned
+}
+
```
I think these would be more at home in `command` or at most in `filetools.go` in the `tools` package (but really they're quite specifically about command line input).

sinbad commented:
👍 except for packaging comments

sinbad commented:
Oh, should have a test too.

ttaylorr commented:
@sinbad great points. I was definitely feeling the weight of `*config.Configuration`. I refactored according to your comments, and added tests as appropriate. Let me know your thoughts!

ttaylorr commented:
Looks like the `prune` tests are still flaky. Will investigate tomorrow. :/

sinbad commented:
Sorry, I should have been more specific: there's a test for the `--include` options in `test-fetch.sh`, you could just add a dir there and prove that using `--include="dir/"` works after your changes.

sinbad commented on config/config.go:
```diff
@@ -76,6 +78,29 @@ func NewConfig() *Configuration {
+//
+// NOTE: this method should only be called during testing.
+func NewFromValues(gitconfig map[string]string) *Configuration {
+	config := &Configuration{
```
You could actually put this in config_test.go to enforce the 'only testing' requirement.

ttaylorr replied:
Originally this `func` was in the `config_test.go` file, but I needed to use it in the [`commands_test.go`](https://github.com/github/git-lfs/pull/1257/files#diff-b65ac838606a4ab61fdb7724d458246dR11). This is sort of smelly to me, but I'm not quite sure what to do about it.

sinbad commented:
Looks good to me, the prune failure is puzzling (auth error), I re-ran the test to be sure it wasn't just an intermittent thing but it happened again.

ttaylorr commented:
@sinbad yeah, not sure what's up with the prune failure, either. @technoweenie and I re-ran the tests in #1267 and the failures were intermittent, so you may have just gotten unlucky.

Thoughts on moving OS X `git-latest-go-1.5` test to the "allowed failures" section?
